### PR TITLE
[EP 228] iOS: Page Viewed (add_ons)

### PR DESCRIPTION
**Pull request by moderateepheezy**
# 📲 What

1. Update `Add-Ons Page Viewed` event name to `Page Viewed`
2. Add CheckoutProperties to `trackAddOnsPageViewed`

# 🤔 Why

Segment Integration

### CODE REVIEW

File: Library/Tracking/KSRAnalytics.swift [hunk 1]
```diff
@@ -733,14 +732,16 @@ public final class KSRAnalytics {
   public func trackAddOnsPageViewed(
     project: Project,
     reward: Reward,
-    refTag: RefTag?
+    refTag: RefTag?,
+    checkoutData: CheckoutPropertiesData
```
Comment (singhhari) [inline Library/Tracking/KSRAnalytics.swift:736]: I've noticed on all the other tracking methods, the `refTag` is the last parameter. Can you update this one as well?

File: Library/Tracking/KSRAnalyticsTests.swift [hunk 2]
```diff
@@ -1468,6 +1468,13 @@ final class KSRAnalyticsTests: TestCase {
     XCTAssertEqual("new_pledge", dataLakeClient.properties.last?["context_type"] as? String)
     XCTAssertEqual("new_pledge", segmentClient.properties.last?["context_type"] as? String)
 
+    ksrAnalytics
+      .trackAddOnsPageViewed(project: .template, reward: .template, refTag: nil, checkoutData: .template)
+    XCTAssertEqual("add_ons", dataLakeClient.properties.last?["context_page"] as? String)
+    XCTAssertEqual("add_ons", segmentClient.properties.last?["context_page"] as? String)
+    XCTAssertEqual(nil, dataLakeClient.properties.last?["context_type"] as? String)
```
Comment (singhhari) [inline Library/Tracking/KSRAnalyticsTests.swift:1475]: No need to check for nil here

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift [hunk 3]
```diff
@@ -455,3 +460,36 @@ private func isValid(
...
+  let addOnsMinimumUsd = Format.decimalCurrency(
+    for: rewards
+      .reduce(0.0) { accum, addOn in accum.addingCurrency(addOn.minimum) }
+      .multiplyingCurrency(staticUsdRate)
+  )
+  return KSRAnalytics.CheckoutPropertiesData(
+    addOnsCountTotal: nil,
+    addOnsCountUnique: nil,
+    addOnsMinimumUsd: addOnsMinimumUsd,
+    amount: "",
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModel.swift:480]: Is there a way we can avoid passing empty strings? Some of the other information available in this view model should make this possible.

File: Library/ViewModels/RewardAddOnSelectionViewModelTests.swift [hunk 4]
```diff
@@ -1107,8 +1107,8 @@ final class RewardAddOnSelectionViewModelTests: TestCase {
     self.vm.inputs.configure(with: data)
     self.vm.inputs.viewDidLoad()
 
-    XCTAssertEqual(["Add-Ons Page Viewed"], self.dataLakeTrackingClient.events)
-    XCTAssertEqual(["Add-Ons Page Viewed"], self.segmentTrackingClient.events)
+    XCTAssertEqual([], self.dataLakeTrackingClient.events)
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModelTests.swift:1110]: Based on the older line this should contain a `Page Viewed` event?

File: Library/ViewModels/RewardAddOnSelectionViewModelTests.swift [hunk 5]
```diff
@@ -1199,8 +1199,8 @@ final class RewardAddOnSelectionViewModelTests: TestCase {
...
+    XCTAssertEqual([], self.segmentTrackingClient.events)
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModelTests.swift:1203]: Same here.

**Review by singhhari — CHANGES_REQUESTED** [review]

**Comment by nativeksr** [comment]
### SwiftFormat found issues:

| File | Rules |
| ---- | ----- |
| Library/ViewModels/RewardAddOnSelectionViewModel.swift | indent |
| Library/ViewModels/RewardAddOnSelectionViewModelTests.swift | indent |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift [hunk 6]
```diff
@@ -256,15 +256,55 @@ public final class RewardAddOnSelectionViewModel: RewardAddOnSelectionViewModelT
...
         AppEnvironment.current.ksrAnalytics.trackAddOnsPageViewed(
           project: project,
           reward: baseReward,
+          checkoutData: checkoutProperties(
+            from: project,
+            baseReward: baseReward,
+            rewards: selectedRewards,
+            selectedQuantities: configData.selectedQuantities,
+            additionalPledgeAmount: 0,
+            pledgeTotal: 0,
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModel.swift:304]: Definitely not 0. We should be using the value of the reward for the total here.

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift [hunk 7]
```diff
@@ -256,15 +256,55 @@ public final class RewardAddOnSelectionViewModel: RewardAddOnSelectionViewModelT
...
+      .observeValues { project, baseReward, selectedRewards, refTag, configData, rewardsShippingTotal in
         AppEnvironment.current.ksrAnalytics.trackAddOnsPageViewed(
           project: project,
           reward: baseReward,
+          checkoutData: checkoutProperties(
+            from: project,
+            baseReward: baseReward,
+            rewards: selectedRewards,
+            selectedQuantities: configData.selectedQuantities,
+            additionalPledgeAmount: 0,
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModel.swift:303]: Depending on if Laurie wants this event to handle an edited pledge this may require you to use the existing `Backing` (project.personalization.backing) instead of hard-coding 0.

File: Library/ViewModels/RewardAddOnSelectionViewModelTests.swift [hunk 8]
```diff
@@ -1107,25 +1107,25 @@ final class RewardAddOnSelectionViewModelTests: TestCase {
...
     self.vm.inputs.viewDidLoad()
 
-    XCTAssertEqual(["Add-Ons Page Viewed"], self.dataLakeTrackingClient.events)
-    XCTAssertEqual(["Add-Ons Page Viewed"], self.segmentTrackingClient.events)
-
     self.scheduler.advance()
     self.vm.inputs.shippingRuleSelected(shippingRule)
     self.scheduler.advance()
 
+    XCTAssertEqual(["Page Viewed"], self.dataLakeTrackingClient.events)
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModelTests.swift:1114]: We shouldn't be moving the line down. We expect the event to be fired after viewDidLoad.

Comment (singhhari) [reply Library/ViewModels/RewardAddOnSelectionViewModelTests.swift:1114]: Ah and can you include a check for the properties after we check the event?

**Review by singhhari — CHANGES_REQUESTED** [review]

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift [hunk 9]
```diff
@@ -256,15 +256,55 @@ public final class RewardAddOnSelectionViewModel: RewardAddOnSelectionViewModelT
     .map(PledgeViewData.init)
     .takeWhen(self.continueButtonTappedProperty.signal)
 
+    let defaultShippingTotal = Signal.zip(project, baseReward)
```
[the remark this answers is not in the document]
Comment (moderateepheezy) [reply Library/ViewModels/RewardAddOnSelectionViewModel.swift:259]: The refactor would result to duplicate we are trying to avoid, because `PledgeViewModel` uses the result of it's calculation else where in the code.

File: Library/Tracking/KSRAnalytics.swift [hunk 10]
```diff
@@ -725,14 +726,16 @@ public final class KSRAnalytics {
   public func trackAddOnsPageViewed(
     project: Project,
     reward: Reward,
+    checkoutData: CheckoutPropertiesData,
     refTag: RefTag?
   ) {
     let props = projectProperties(from: project, loggedInUser: self.loggedInUser)
       .withAllValuesFrom(pledgeProperties(from: reward))
```
[the remark this answers is not in the document]
Comment (moderateepheezy) [reply Library/Tracking/KSRAnalytics.swift:733]: Will correct that now.

File: Library/Tracking/KSRAnalytics.swift [hunk 11]
```diff
@@ -26,6 +26,7 @@ public final class KSRAnalytics {
     case activityFeedViewed = "Activity Feed Viewed"
     case addNewCardButtonClicked = "Add New Card Button Clicked"
     case addOnsContinueButtonClicked = "Add-Ons Continue Button Clicked"
+    case campaignDetailsButtonClicked = "Campaign Details Button Clicked"
```
[the remark this answers is not in the document]
Comment (moderateepheezy) [reply Library/Tracking/KSRAnalytics.swift:29]: I didn't add this, this is probably from a merge.

[the remark this answers is not in the document]
Comment (singhhari) [reply Library/Tracking/KSRAnalytics.swift:29]: Cool, after you pull the changes it should be removed automatically.

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (singhhari) [reply Library/ViewModels/RewardAddOnSelectionViewModel.swift:259]: Not sure I understand what you mean. We don't need to replicate the logic of the PledgeViewModel everywhere. My understanding is that the shipping, at this point, should only be the shipping cost associated with the base reward since the add on has not been selected yet. But again, clarify with Laurie about whether this event should handle only new pledges or both new pledges + edits.

File: Library/ViewModels/RewardAddOnSelectionViewModelTests.swift (hunk 8 above)
Comment (moderateepheezy) [reply Library/ViewModels/RewardAddOnSelectionViewModelTests.swift:1114]: So what fires the "Page Viewed" event is inside an asynchronous task, `XCTAssertEqual(["Page Viewed"], self.dataLakeTrackingClient.events)` is being called immediately after viewDidLoad, so there's a possibility that `XCTAssertEqual(["Page Viewed"], self.dataLakeTrackingClient.events)` will be called before the "Page Viewed" event is fired. 

Moving `XCTAssertEqual(["Page Viewed"], self.dataLakeTrackingClient.events)` to after `self.scheduler.advance()` would ensure that all actions that have been scheduled are executed  as soon as possible.

 I think the fact that it passes previously is a false positive.

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift (hunk 7 above)
Comment (moderateepheezy) [reply Library/ViewModels/RewardAddOnSelectionViewModel.swift:276]: This has updated as well.

File: Library/ViewModels/RewardAddOnSelectionViewModelTests.swift (hunk 8 above)
Comment (singhhari) [reply Library/ViewModels/RewardAddOnSelectionViewModelTests.swift:1114]: Not sure I agree with this. If the test was working before then it is fair to assume it was correct. The issue is very likely related to the changes you applied to the view model. Keep in mind many more parameters were introduced to the tracking logic (`RewardAddOnSelectionViewModel line 284`) and that is likely the source of the issue.

**Review by singhhari — CHANGES_REQUESTED** [review]

File: Library/SharedFunctions.swift [hunk 12]
```diff
@@ -397,6 +397,88 @@ public func rounded(_ value: Double, places: Int) -> Double {
...
+ An helper func that calculates  shipping total for base reward
+
+ - parameter project: The `Project` associated with a group of Rewards.
+ - parameter baseReward: The reward being evaluated
+
+ - returns: A `Double` of the shipping value. If the `Project` `Backing` object is nil,
+            and `baseReward` shipping is not enabled, the value is `0.0`
+ */
+public func getBaseRewardShippingTotal(project: Project, baseReward: Reward) -> Double {
+  guard baseReward.shipping.enabled, let backing = project.personalization.backing else { return 0.0 }
```
Comment (singhhari) [inline Library/SharedFunctions.swift:410]: So I just realized something while I was testing on your branch. On a new pledge this always return 0 because there is no `Backing`.  The desired behavior is to return the shipping amount designated on the Reward for a new pledge.

I tested this while doing a new pledge for the project `The Tarot Restless: Third Edition`. The shipping amount is $5 to the destination I chose but we are sending `checkout_shipping_amount`: 0 to Segment.

File: Library/SharedFunctionsTests.swift [hunk 13]
```diff
@@ -404,4 +404,116 @@ final class SharedFunctionsTests: TestCase {
       checkoutPropertiesData.userHasStoredApplePayCard
     )
   }
+
+  func testGetShipping_ShippingEnabled() {
```
Comment (singhhari) [inline Library/SharedFunctionsTests.swift:408]: This test appears to be actually testing a situation where you are testing the shipping amount based on an existing Backing so the name is not accurate imo. 

I would consider renaming this something like `testGetShipping_Backing` and writing another test for the situation when the Backing doesn't exist and the Reward has a non-zero shipping amount.

File: Library/SharedFunctionsTests.swift [hunk 14]
```diff
@@ -404,4 +404,116 @@ final class SharedFunctionsTests: TestCase {
...
+      )
+
+    let shippingTotal = getBaseRewardShippingTotal(project: project, baseReward: reward)
+
+    XCTAssertEqual(2.0, shippingTotal)
+  }
+
+  func testGetShipping_ShippingDisabled() {
+    let reward = Reward.template
+      |> Reward.lens.limit .~ 5
```
Comment (singhhari) [inline Library/SharedFunctionsTests.swift:432]: The test does not require 432-434

File: Library/SharedFunctionsTests.swift [hunk 15]
```diff
@@ -404,4 +404,116 @@ final class SharedFunctionsTests: TestCase {
...
+    let shippingTotal = calculateShippingTotal(
+      shippingRule: shippingRule,
+      addOnRewards: [reward],
+      quantities: [:]
+    )
+
+    XCTAssertEqual(0.0, shippingTotal)
+  }
+
+  func testCalculated_Pledge_Total() {
```
Comment (singhhari) [inline Library/SharedFunctionsTests.swift:489]: Can we try to test this by combining the outputs of the other functions you defined instead of hardcoding the arguments?

File: Library/ViewModels/RewardAddOnSelectionViewModelTests.swift [hunk 16]
```diff
@@ -1107,25 +1107,27 @@ final class RewardAddOnSelectionViewModelTests: TestCase {
     self.vm.inputs.configure(with: data)
     self.vm.inputs.viewDidLoad()
 
-    XCTAssertEqual(["Add-Ons Page Viewed"], self.dataLakeTrackingClient.events)
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModelTests.swift:1110]: I remember we spoke about this last week. This should not have been deleted and moved down. Please look into this as I'm fairly certain it is the designed functionality. It is probably directly related to the changes applied to the `RewardAddOnSelectionViewModel`.

File: Library/ViewModels/RewardAddOnSelectionViewModelTests.swift [hunk 17]
```diff
@@ -1199,13 +1201,15 @@ final class RewardAddOnSelectionViewModelTests: TestCase {
...
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModelTests.swift:1202]: Same here.

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift [hunk 18]
```diff
@@ -256,18 +256,79 @@ public final class RewardAddOnSelectionViewModel: RewardAddOnSelectionViewModelT
...
     .takeWhen(self.continueButtonTappedProperty.signal)
 
+    let calculatedShippingTotal = Signal.combineLatest(
+      shippingRule.skipNil(),
+      allRewards,
+      selectedQuantities
+    )
+    .map(calculateShippingTotal)
+
+    let shippingTotal = Signal.zip(project, baseReward)
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModel.swift:266]: This var should probably be called `baseRewardShippingTotal` based on the name of the method being used.

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift [hunk 19]
```diff
@@ -256,18 +256,79 @@ public final class RewardAddOnSelectionViewModel: RewardAddOnSelectionViewModelT
     .map(PledgeViewData.init)
     .takeWhen(self.continueButtonTappedProperty.signal)
 
+    let calculatedShippingTotal = Signal.combineLatest(
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModel.swift:259]: This should be `shippingTotal` based on the method name

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift [hunk 20]
```diff
@@ -256,18 +256,79 @@ public final class RewardAddOnSelectionViewModel: RewardAddOnSelectionViewModelT
...
+    )
+    .take(first: 1)
+
+    let allRewardsTotal = Signal.combineLatest(
+      selectedRewards,
+      selectedQuantities
+    )
+    .map(calculateAllRewardsTotal)
+
+    let calculatedPledgeTotal = Signal.combineLatest(
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModel.swift:287]: This is an opinion but the variable name here is also not very descriptive. It should probably be something like `combinedPledgeTotal` or something along those lines.

**Review by singhhari — CHANGES_REQUESTED** [review]

File: Library/SharedFunctions.swift [hunk 21]
```diff
@@ -397,6 +397,99 @@ public func rounded(_ value: Double, places: Int) -> Double {
...
+  }
+
+  return r
+}
+
+/**
+ An helper func that calculates  pledge total for all rewards
+
+ - parameter addOnRewards: The `Project` associated with a group of Rewards.
+ - parameter selectedQuantities: The shipping cost for the pledge.
```
Comment (singhhari) [inline Library/SharedFunctions.swift:478]: Update description of the parameter

File: Library/SharedFunctionsTests.swift [hunk 22]
```diff
@@ -404,4 +404,145 @@ final class SharedFunctionsTests: TestCase {
...
+      baseReward: reward,
+      shippingRule: shippingRule
+    )
+
+    XCTAssertEqual(5.0, shippingTotal)
+  }
+
+  func testGetShipping_ShippingDisabled() {
+    let reward = Reward.template
+      |> Reward.lens.limit .~ 5
```
Comment (singhhari) [inline Library/SharedFunctionsTests.swift:455]: 455-457 not needed here

File: Library/SharedFunctionsTests.swift [hunk 23]
```diff
@@ -404,4 +404,145 @@ final class SharedFunctionsTests: TestCase {
...
+
+    let quantities: SelectedRewardQuantities = [
+      reward.id: 1,
+      addOn1.id: 2,
+      addOn2.id: 1
+    ]
+
+    let shippingTotal = calculateShippingTotal(
+      shippingRule: shippingRule,
+      addOnRewards: [reward],
```
Comment (singhhari) [inline Library/SharedFunctionsTests.swift:495]: Shouldn't this contain the addOn1 and addOn2 defined above?

File: Library/SharedFunctionsTests.swift [hunk 24]
```diff
@@ -404,4 +404,145 @@ final class SharedFunctionsTests: TestCase {
...
+      quantities: [:]
+    )
+
+    XCTAssertEqual(0.0, shippingTotal)
+  }
+
+  func testCalculated_Pledge_Total() {
```
Comment (singhhari) [inline Library/SharedFunctionsTests.swift:518]: It would ideal if you tested this by defining the rewards/addons and using their amounts to calculate the total

File: Library/SharedFunctionsTests.swift [hunk 25]
```diff
@@ -404,4 +404,145 @@ final class SharedFunctionsTests: TestCase {
...
+    let addOn2 = Reward.template
+      |> Reward.lens.id .~ 10
+
+    let quantities: SelectedRewardQuantities = [
+      reward.id: 1,
+      addOn1.id: 2,
+      addOn2.id: 1
+    ]
+
+    let rewardsTotal = calculateAllRewardsTotal(
```
Comment (singhhari) [inline Library/SharedFunctionsTests.swift:541]: Shouldnt selectedQuantities and addOnRewards be related here? I'm seeing the quantities with the reward and addons but the addOnRewards argument is just an array with one reward in it.

File: Library/ViewModels/PledgeViewModel.swift [hunk 26]
```diff
@@ -156,40 +156,17 @@ public class PledgeViewModel: PledgeViewModelType, PledgeViewModelInputs, Pledge
...
-    }
-
-    let defaultShippingTotal = Signal.zip(project, baseReward)
-      .map { project, baseReward -> Double in
-        guard baseReward.shipping.enabled, let backing = project.personalization.backing else { return 0.0 }
+    .map(calculateShippingTotal)
 
-        return backing.shippingAmount.flatMap(Double.init) ?? 0.0
-      }
+    let defaultShippingTotal = Signal.zip(project, baseReward, selectedShippingRule)
```
Comment (singhhari) [inline Library/ViewModels/PledgeViewModel.swift:168]: a better name would be related to the method being applied. something like `baseRewardShippingTotal`

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift [hunk 27]
```diff
@@ -256,18 +256,80 @@ public final class RewardAddOnSelectionViewModel: RewardAddOnSelectionViewModelT
...
+      baseRewardShippingTotal,
+      shippingTotal
+    )
+
+    // Addtional pledge amount is zero if not backed.
+    let additionalPledgeAmount = Signal.merge(
+      configData.filter { $0.project.personalization.backing == nil }.mapConst(0.0),
+      project.map { $0.personalization.backing }.skipNil().map(\.bonusAmount)
+    )
+    .take(first: 1)
```
Comment (singhhari) [inline Library/ViewModels/RewardAddOnSelectionViewModel.swift:279]: Curious why take first here?

File: Library/ViewModels/RewardAddOnSelectionViewModelTests.swift (hunk 16 above)
Comment (singhhari) [reply Library/ViewModels/RewardAddOnSelectionViewModelTests.swift:1110]: We determined that because the shipping rule is required before the event is fired, it makes sense that this assertion is moved lower.

**Review by singhhari — CHANGES_REQUESTED** [review]

File: Library/SharedFunctionsTests.swift [hunk 28]
```diff
@@ -404,4 +404,148 @@ final class SharedFunctionsTests: TestCase {
...
+      |> Reward.lens.id .~ 5
+    let addOn2 = Reward.template
+      |> Reward.lens.id .~ 10
+
+    let shippingRule = ShippingRule.template
+
+    let shippingTotal = calculateShippingTotal(
+      shippingRule: shippingRule,
+      addOnRewards: [addOn1, addOn2],
+      quantities: [:]
```
Comment (singhhari) [inline Library/SharedFunctionsTests.swift:515]: supply the quantities based on the addons and reward defined

**Review by singhhari — CHANGES_REQUESTED** [review]
Whatever is unresolved is the only bit left and then we can merge this,.

File: Library/SharedFunctionsTests.swift [hunk 29]
```diff
@@ -509,21 +509,113 @@ final class SharedFunctionsTests: TestCase {
...
+    )
+
+    let allRewardsTotal = calculateAllRewardsTotal(
+      addOnRewards: addOnRewards,
+      selectedQuantities: quantities
+    )
+
     let pledgeTotal = calculatePledgeTotal(
-      pledgeAmount: 20.0, shippingCost: 5.0, addOnRewardsTotal: 50
+      pledgeAmount: 0.0,
```
[the remark this answers is not in the document]
Comment (moderateepheezy) [reply Library/SharedFunctionsTests.swift]: Done.

File: Library/SharedFunctionsTests.swift [hunk 30]
```diff
@@ -509,21 +509,113 @@ final class SharedFunctionsTests: TestCase {
...
+      |> Reward.lens.hasAddOns .~ true
+      |> Reward.lens.shipping.enabled .~ true
+    let addOn1 = Reward.template
+      |> Reward.lens.id .~ 5
+      |> Reward.lens.shipping .~ (.template |> Reward.Shipping.lens.enabled .~ true)
+    let addOn2 = Reward.template
+      |> Reward.lens.id .~ 10
+      |> Reward.lens.shipping .~ (.template |> Reward.Shipping.lens.enabled .~ true)
+
+    let shippingRule = ShippingRule.template
```
[the remark this answers is not in the document]
Comment (moderateepheezy) [reply Library/SharedFunctionsTests.swift]: Done

File: Library/SharedFunctionsTests.swift [hunk 31]
```diff
@@ -509,21 +509,113 @@ final class SharedFunctionsTests: TestCase {
...
+      reward.id: 1,
+      addOn1.id: 2,
+      addOn2.id: 1
+    ]
+
     let shippingTotal = calculateShippingTotal(
       shippingRule: shippingRule,
-      addOnRewards: [addOn1, addOn2],
-      quantities: [:]
+      addOnRewards: [reward, addOn1, addOn2],
```
[the remark this answers is not in the document]
Comment (moderateepheezy) [reply Library/SharedFunctionsTests.swift]: Done

File: Library/SharedFunctionsTests.swift [hunk 32]
```diff
@@ -509,21 +509,113 @@ final class SharedFunctionsTests: TestCase {
...
+      |> Project.lens.personalization.backing .~ (
+        .template
+          |> Backing.lens.reward .~ reward
+          |> Backing.lens.rewardId .~ reward.id
+          |> Backing.lens.shippingAmount .~ 10
+          |> Backing.lens.amount .~ 700.0
+          |> Backing.lens.bonusAmount .~ 700.0
+      )
+
+    let shippingRule = ShippingRule.template
```
[the remark this answers is not in the document]
Comment (moderateepheezy) [reply Library/SharedFunctionsTests.swift]: Done

**Review by singhhari — APPROVED** [review]
lgtm
